src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (C) 2003-2015 The Music Player Daemon Project
+ * http://www.musicpd.org
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+#include "UsfDecoderPlugin.hxx"
+#include "../DecoderAPI.hxx"
+#include "../DecoderInternal.hxx"
+#include "tag/TagHandler.hxx"
+#include "fs/Path.hxx"
+#include "util/Domain.hxx"
+#include "Log.hxx"
+
+#include <usf.h>
+#include <psflib.h>
+
+static constexpr Domain usf_domain("usf");
+static constexpr unsigned USF_BUFFER_FRAMES = 2048;
+static constexpr unsigned USF_CHANNELS = 2;
+static constexpr unsigned USF_BUFFER_SAMPLES = USF_BUFFER_FRAMES*USF_CHANNELS;
+
+static void * stdio_fopen( const char * path )
+{
+    return fopen( path, "rb" );
+}
+
+static size_t stdio_fread( void *p, size_t size, size_t count, void *f )
+{
+    return fread( p, size, count, (FILE*) f );
+}
+
+static int stdio_fseek( void * f, int64_t offset, int whence )
+{
+    return fseek( (FILE*) f, offset, whence );
+}
+
+static int stdio_fclose( void * f )
+{
+    return fclose( (FILE*) f );
+}
+
+static long stdio_ftell( void * f )
+{
+    return ftell( (FILE*) f );
+}
+
+psf_file_callbacks stdio_callbacks =
+{
+    "\\/:",
+    stdio_fopen,
+    stdio_fread,
+    stdio_fseek,
+    stdio_fclose,
+    stdio_ftell
+};
+
+struct usf_loader_state
```

Comment:
The variables are passed into the third party library's usf_set_compare and usf_set_fifo_full functions. I didn't write the libraries responsible for the actual decoding. In this plugin they are merely read from the tags of the input file and passed through to the emulator.